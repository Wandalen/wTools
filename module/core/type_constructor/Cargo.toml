[package]
name = "type_constructor"
version = "0.2.0"
edition = "2021"
authors = [
  "Kostiantyn Wandalen <wandalen@obox.systems>",
  "Dmytro Kryvoruchko <dmytro.kr@obox.systems>",
]
license = "MIT"
readme = "Readme.md"
documentation = "https://docs.rs/type_constructor"
repository = "https://github.com/Wandalen/wTools/tree/master/module/core/type_constructor"
homepage = "https://github.com/Wandalen/wTools/tree/master/module/core/type_constructor"
description = """
Fundamental data types and type constructors, like Single, Pair, Many.
"""
categories = [ "algorithms", "development-tools" ]
keywords = [ "fundamental", "general-purpose" ]

[lints]
workspace = true

[package.metadata.docs.rs]
features = [ "full" ]
all-features = false

exclude = [ "/tests", "/examples", "-*" ]
include = [
  "/rust/impl/dt/type_constructor",
  "/Cargo.toml",
  "/Readme.md",
  "/License",
]

[features]
default = [
  "enabled",
  "many",
  "make",
  "vectorized_from",
]
full = [
  "enabled",
  "use_alloc",
  "many",
  "make",
  "vectorized_from",
]
no_std = []
use_alloc = []
enabled = []
many = []
make = []
vectorized_from = []

# [lib]
# name = "type_constructor"
# path = "src/dt/type_constructor/type_constructor_lib.rs"
#
# [[test]]
# name = "type_constructor_test"
# path = "tests/dt/type_constructor/type_constructor_tests.rs"
#
# [[test]]
# name = "type_constructor_smoke_test"
# path = "tests/_integration_test/smoke_test.rs"

# [[example]]
# name = "type_constructor_trivial_sample"
# path = "examples/type_constructor_trivial_sample/src/main.rs"
#
# [[example]]
# name = "type_constructor_derive_and_attr_sample"
# path = "examples/type_constructor_derive_and_attr_sample/src/main.rs"
#
# [[example]]
# name = "type_constructor_struct_sample"
# path = "examples/type_constructor_struct_sample/src/main.rs"
#
# [[example]]
# name = "type_constructor_parametrized_element_sample"
# path = "examples/type_constructor_parametrized_element_sample/src/main.rs"
#
# [[example]]
# name = "type_constructor_parametrized_tuple_sample"
# path = "examples/type_constructor_parametrized_tuple_sample/src/main.rs"
#
# [[example]]
# name = "type_constructor_multiple_sample"
# path = "examples/type_constructor_multiple_sample/src/main.rs"
# required-features = [ "many" ]
#
# [[example]]
# name = "type_constructor_without_macro_sample"
# path = "examples/type_constructor_without_macro_sample/src/main.rs"
# required-features = [ "many" ]
#
# [[example]]
# name = "type_constructor_pair_sample"
# path = "examples/type_constructor_pair_sample/src/main.rs"
#
# [[example]]
# name = "type_constructor_homopair_sample"
# path = "examples/type_constructor_homopair_sample/src/main.rs"
#
# [[example]]
# name = "type_constructor_many_sample"
# path = "examples/type_constructor_many_sample/src/main.rs"
# required-features = [ "many" ]

[dependencies]
derive_tools = { workspace = true, features = [ "enabled", "type_variadic_from", "derive_variadic_from" ] }
# inspect_type = { workspace = true }
# type_constructor_make_meta = { workspace = true }
# type_constructor_derive_pair_meta = { workspace = true }

[dev-dependencies]
test_tools = { workspace = true }
