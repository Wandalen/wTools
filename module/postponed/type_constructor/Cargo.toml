[package]
name = "type_constructor"
version = "0.3.0"
edition = "2021"
authors = [
  "Kostiantyn Wandalen <wandalen@obox.systems>",
  "Dmytro Kryvoruchko <dmytro.kr@obox.systems>",
]
license = "MIT"
readme = "Readme.md"
documentation = "https://docs.rs/type_constructor"
repository = "https://github.com/Wandalen/wTools/tree/master/module/core/type_constructor"
homepage = "https://github.com/Wandalen/wTools/tree/master/module/core/type_constructor"
description = """
Fundamental data types and type constructors, like Single, Pair, Many.
"""
categories = [ "algorithms", "development-tools" ]
keywords = [ "fundamental", "general-purpose" ]

[lints]
workspace = true

[package.metadata.docs.rs]
features = [ "full" ]
all-features = false


include = [
  "/rust/impl/dt/type_constructor",
  "/Cargo.toml",
  "/Readme.md",
  "/License",
]

[features]
default = [
  "enabled",
  "many",
  "make",
  "vectorized_from",
]
full = [
  "enabled",
  "use_alloc",
  "many",
  "make",
  "vectorized_from",
]
no_std = []
use_alloc = [ "no_std" ]
enabled = []
many = []
make = []
vectorized_from = []

# [lib]
# name = "type_constructor"
# path = "src/dt/type_constructor/type_constructor_lib.rs"
#
# [[test]]
# name = "type_constructor_test"
# path = "tests/dt/type_constructor/type_constructor_tests.rs"
#
# [[test]]
# name = "type_constructor_smoke_test"
# path = "tests/_integration_test/smoke_test.rs"

# [[example]]
# name = "type_constructor_trivial"
# path = "examples/type_constructor_trivial/src/main.rs"
#
# [[example]]
# name = "type_constructor_derive_and_attr"
# path = "examples/type_constructor_derive_and_attr/src/main.rs"
#
# [[example]]
# name = "type_constructor_struct"
# path = "examples/type_constructor_struct/src/main.rs"
#
# [[example]]
# name = "type_constructor_parametrized_element"
# path = "examples/type_constructor_parametrized_element/src/main.rs"
#
# [[example]]
# name = "type_constructor_parametrized_tuple"
# path = "examples/type_constructor_parametrized_tuple/src/main.rs"
#
# [[example]]
# name = "type_constructor_multiple"
# path = "examples/type_constructor_multiple/src/main.rs"
# required-features = [ "many" ]
#
# [[example]]
# name = "type_constructor_without_macro"
# path = "examples/type_constructor_without_macro/src/main.rs"
# required-features = [ "many" ]
#
# [[example]]
# name = "type_constructor_pair"
# path = "examples/type_constructor_pair/src/main.rs"
#
# [[example]]
# name = "type_constructor_homopair"
# path = "examples/type_constructor_homopair/src/main.rs"
#
# [[example]]
# name = "type_constructor_many"
# path = "examples/type_constructor_many/src/main.rs"
# required-features = [ "many" ]

[dependencies]
derive_tools = { workspace = true, features = [ "enabled", "type_variadic_from", "derive_variadic_from" ] }
# inspect_type = { workspace = true }
# type_constructor_make_meta = { workspace = true }
# type_constructor_derive_pair_meta = { workspace = true }

[dev-dependencies]
test_tools = { workspace = true }
