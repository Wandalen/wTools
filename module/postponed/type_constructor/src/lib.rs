
#![ cfg_attr( feature = "no_std", no_std ) ]
#![ doc( html_logo_url = "https://raw.githubusercontent.com/Wandalen/wTools/master/asset/img/logo_v3_trans_square.png" ) ]
#![ doc( html_favicon_url = "https://raw.githubusercontent.com/Wandalen/wTools/alpha/asset/img/logo_v3_trans_square_icon_small_v2.ico")]
#![ doc( html_root_url = "https://docs.rs/type_constructor/latest/type_constructor/")]
// #![ deny( rust_2018_idioms ) ]
// #![ deny( missing_debug_implementations ) ]
// #![ deny( missing_docs ) ]

//!
//! Type constructors of fundamental data types.
//!

#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/", "Readme.md" ) ) ]

pub use derive_tools::{ From_0, From_1, From_2, From_3, from };

/// Temporary workaround.
#[ macro_export ]
macro_rules! _if_from
{
  ( $( $code:tt )* )
  =>
  {
    $( $code )*
  };
}

// #![ without_std ]

// #[ cfg( feature = "no_std" ) ]
// extern crate core as std;
// #[ cfg( all( feature = "no_std", feature = "use_alloc" ) ) ]
// extern crate alloc;

// #[ path = "./inc.rs" ]
// mod inc;
// pub mod type_constuctor;
// #[ doc( inline ) ]
// #[ allow( unused_imports ) ]
// pub use inc::*;


#[ cfg( feature = "enabled" ) ]
pub mod type_constuctor;

/// Namespace with dependencies.

#[ cfg( feature = "enabled" ) ]
pub mod dependency
{
}

#[ doc( inline ) ]
#[ allow( unused_imports ) ]
#[ cfg( feature = "enabled" ) ]
pub use own::*;

/// Own namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod own
{
  use super::*;
  #[ doc( inline ) ]
  #[ allow( unused_imports ) ]
  pub use orphan::*;
  pub use super::type_constuctor::own::*;
}

/// Shared with parent namespace of the module
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod orphan
{
  use super::*;
  #[ doc( inline ) ]
  #[ allow( unused_imports ) ]
  pub use exposed::*;
  pub use super::type_constuctor::orphan::*;
}

/// Exposed namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod exposed
{
  use super::*;
  #[ doc( inline ) ]
  #[ allow( unused_imports ) ]
  pub use prelude::*;
  #[ doc( inline ) ]
  #[ allow( unused_imports ) ]
  pub use super::type_constuctor::exposed::*;
}

/// Prelude to use essentials: `use my_module::prelude::*`.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod prelude
{
  use super::*;
  pub use super::type_constuctor::prelude::*;
}
