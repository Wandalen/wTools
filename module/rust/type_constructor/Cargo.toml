[package]
name = "type_constructor"
version = "0.1.20"
edition = "2021"
authors = [
  "Kostiantyn Wandalen <wandalen@obox.systems>",
  "Dmytro Kryvoruchko <dm.vict.kr@gmail.com>",
]
license = "MIT"
readme = "Readme.md"
documentation = "https://docs.rs/type_constructor"
repository = "https://github.com/Wandalen/wTools/tree/master/module/rust/type_constructor"
homepage = "https://github.com/Wandalen/wTools/tree/master/module/rust/type_constructor"
description = """
Fundamental data types and type constructors, like Single, Pair, Many.
"""
categories = [ "algorithms", "development-tools" ]
keywords = [ "fundamental", "general-purpose" ]

[package.metadata.docs.rs]
all-features = true

include = [
  "/rust/impl/dt/type_constructor",
  "/Cargo.toml",
  "/Readme.md",
  "/License",
]

[features]
default = [
  "use_std",
  "many",
  "make",
  "vectorized_from",
]
full = [
  "use_std",
  "use_alloc",
  "many",
  "make",
  "vectorized_from",
]
use_std = []
use_alloc = []

many = []
make = []
vectorized_from = []

[lib]
name = "type_constructor"
path = "rust/impl/dt/type_constructor/type_constructor_lib.rs"

[[test]]
name = "type_constructor_test"
path = "rust/test/dt/type_constructor/type_constructor_tests.rs"

[[test]]
name = "type_constructor_smoke_test"
path = "rust/test/_integration_test/smoke_test.rs"

[[example]]
name = "type_constructor_trivial_sample"
path = "sample/rust/type_constructor_trivial_sample/src/main.rs"

[[example]]
name = "type_constructor_derive_and_attr_sample"
path = "sample/rust/type_constructor_derive_and_attr_sample/src/main.rs"

[[example]]
name = "type_constructor_struct_sample"
path = "sample/rust/type_constructor_struct_sample/src/main.rs"

[[example]]
name = "type_constructor_parametrized_element_sample"
path = "sample/rust/type_constructor_parametrized_element_sample/src/main.rs"

[[example]]
name = "type_constructor_parametrized_tuple_sample"
path = "sample/rust/type_constructor_parametrized_tuple_sample/src/main.rs"

[[example]]
name = "type_constructor_multiple_sample"
path = "sample/rust/type_constructor_multiple_sample/src/main.rs"
required-features = [ "many" ]

[[example]]
name = "type_constructor_without_macro_sample"
path = "sample/rust/type_constructor_without_macro_sample/src/main.rs"
required-features = [ "many" ]

[[example]]
name = "type_constructor_pair_sample"
path = "sample/rust/type_constructor_pair_sample/src/main.rs"

[[example]]
name = "type_constructor_homopair_sample"
path = "sample/rust/type_constructor_homopair_sample/src/main.rs"

[[example]]
name = "type_constructor_many_sample"
path = "sample/rust/type_constructor_many_sample/src/main.rs"
required-features = [ "many" ]

[dependencies]
# inspect_type = { version = "~0.1", path = "../../rust/inspect_type", features = [ "nightly", "use_std" ] }

[dev-dependencies]
test_tools = { version = "~0.1", path = "../../rust/test_tools" }
