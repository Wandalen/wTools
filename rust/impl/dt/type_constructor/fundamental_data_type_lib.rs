#![ doc( html_logo_url = "https://raw.githubusercontent.com/Wandalen/wTools/master/asset/img/logo_v3_trans_square.png" ) ]
#![ warn( rust_2018_idioms ) ]
#![ warn( missing_debug_implementations ) ]
#![ warn( missing_docs ) ]

//!
//! Fundamental data types and type constructors, like Single, Pair, Many.
//!

#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/Readme.md" ) ) ]

#[ doc( inline ) ]
pub use ::type_constructor::*;
