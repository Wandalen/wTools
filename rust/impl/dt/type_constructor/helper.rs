/// Internal namespace.
pub( crate ) mod private
{
  use crate::exposed::*;

  ///
  /// Generate code only if feature::make is enabled.
  ///
  /// Do not use manually.
  ///

  #[ cfg( feature = "make" ) ]
  #[ macro_export ]
  macro_rules! _if_make
  {
    ( $( $Rest : tt )* ) =>
    {
      $( $Rest )*
    };
  }

  ///
  /// Generate code only if feature::make is disabled.
  ///
  /// Do not use manually.
  ///

  #[ cfg( not( feature = "make" ) ) ]
  #[ macro_export ]
  macro_rules! _if_make
  {
    ( $( $Rest : tt )* ) =>
    {
    };
  }

  pub use _if_make;
}

/// Protected namespace of the module.
pub mod protected
{
  #[ doc( inline ) ]
  pub use super::orphan::*;
}

#[ doc( inline ) ]
pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  #[ doc( inline ) ]
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  #[ doc( inline ) ]
  pub use super::prelude::*;
}

#[ doc( inline ) ]
pub use exposed::*;

/// Prelude to use essentials: `use my_module::prelude::*`.
pub mod prelude
{
  #[ doc( inline ) ]
  pub use super::private::
  {
    _if_make,
  };
}
