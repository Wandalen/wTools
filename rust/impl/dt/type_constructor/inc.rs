
/// Type constructor of many.
// #[ cfg( feature = "types" ) ]
pub mod many;
/// Type constructor of pair.
// #[ cfg( feature = "types" ) ]
pub mod pair;
/// Type constructor of single.
// #[ cfg( feature = "types" ) ]
pub mod single;
/// Type constructors.
// #[ cfg( feature = "types" ) ]
pub mod types;

/// Macro helpers.
pub mod helper;
/// Generic traits.
pub mod traits;
/// Variadic constructor.
#[ cfg( feature = "make" ) ]
pub mod make;

/// Protected namespace of the module.
pub mod protected
{
  #[ doc( inline ) ]
  pub use super::orphan::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::many::orphan::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::pair::orphan::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::single::orphan::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::types::orphan::*;
  #[ doc( inline ) ]
  pub use super::helper::orphan::*;
  #[ doc( inline ) ]
  pub use super::traits::orphan::*;
  #[ doc( inline ) ]
  #[ cfg( feature = "make" ) ]
  pub use super::make::orphan::*;
}

#[ doc( inline ) ]
pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  #[ doc( inline ) ]
  pub use super::prelude::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::many::exposed::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::pair::exposed::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::single::exposed::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::types::exposed::*;
  #[ doc( inline ) ]
  pub use super::helper::exposed::*;
  #[ doc( inline ) ]
  pub use super::traits::exposed::*;
  #[ doc( inline ) ]
  #[ cfg( feature = "make" ) ]
  pub use super::make::exposed::*;
}

/// Prelude to use essentials: `use my_module::prelude::*`.
pub mod prelude
{
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::many::prelude::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::pair::prelude::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::single::prelude::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::types::prelude::*;
  #[ doc( inline ) ]
  pub use super::helper::prelude::*;
  #[ doc( inline ) ]
  pub use super::traits::prelude::*;
  #[ doc( inline ) ]
  #[ cfg( feature = "make" ) ]
  pub use super::make::prelude::*;
}
