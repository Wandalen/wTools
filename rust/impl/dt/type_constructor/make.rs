/// Internal namespace.
#[ cfg( feature = "make" ) ]
pub( crate ) mod private
{

  ///
  /// Constructor without arguments.
  ///

  pub trait Make0
  where
    Self : Sized,
  {
    /// Constructor without arguments.
    fn make() -> Self
    {
      Self::make_0()
    }
    /// Constructor without arguments.
    fn make_0() -> Self;
  }

  ///
  /// Constructor with single argument.
  ///

  pub trait Make1< Arg >
  where
    Self : Sized,
  {
    /// Constructor without arguments.
    fn make( arg : Arg ) -> Self
    {
      Self::make_1( arg )
    }
    /// Constructor without arguments.
    fn make_1( arg : Arg ) -> Self;
  }

  ///
  /// Constructor with two arguments.
  ///

  pub trait Make2< Arg1, Arg2 >
  where
    Self : Sized,
  {
    /// Constructor with two arguments.
    fn make( arg1 : Arg1, arg2 : Arg2 ) -> Self
    {
      Self::make_2( arg1, arg2 )
    }
    /// Constructor with two arguments.
    fn make_2( arg1 : Arg1, arg2 : Arg2 ) -> Self;
  }

  ///
  /// Constructor with three arguments.
  ///

  pub trait Make3< Arg1, Arg2, Arg3 >
  where
    Self : Sized,
  {
    /// Constructor with three arguments.
    fn make( arg1 : Arg1, arg2 : Arg2, arg3 : Arg3 ) -> Self
    {
      Self::make_3( arg1, arg2, arg3 )
    }
    /// Constructor with three arguments.
    fn make_3( arg1 : Arg1, arg2 : Arg2, arg3 : Arg3 ) -> Self;
  }

  ///
  /// Variadic constructor.
  ///
  /// Implement traits [Make0], [Make1] up to MakeN to provide the interface to construct your structure with a different set of arguments.
  /// In this example structure, Struct1 could be constructed either without arguments, with a single argument, or with two arguments.
  /// - Constructor without arguments fills fields with zero.
  /// - Constructor with a single argument sets both fields to the value of the argument.
  /// - Constructor with 2 arguments set individual values of each field.
  ///
  /// ```rust
  /// #[ cfg( feature = "make" ) ]
  /// {
  ///   use type_constructor::prelude::*;
  ///
  ///   #[ derive( Debug, PartialEq ) ]
  ///   struct Struct1
  ///   {
  ///     a : i32,
  ///     b : i32,
  ///   }
  ///
  ///   impl Make0 for Struct1
  ///   {
  ///     fn make_0() -> Self
  ///     {
  ///       Self { a : 0, b : 0 }
  ///     }
  ///   }
  ///
  ///   impl Make1< i32 > for Struct1
  ///   {
  ///     fn make_1( val : i32 ) -> Self
  ///     {
  ///       Self { a : val, b : val }
  ///     }
  ///   }
  ///
  ///   impl Make2< i32, i32 > for Struct1
  ///   {
  ///     fn make_2( val1 : i32, val2 : i32 ) -> Self
  ///     {
  ///       Self { a : val1, b : val2 }
  ///     }
  ///   }
  ///
  ///   let got : Struct1 = make!();
  ///   let exp = Struct1{ a : 0, b : 0 };
  ///   assert_eq!( got, exp );
  ///
  ///   let got : Struct1 = make!( 13 );
  ///   let exp = Struct1{ a : 13, b : 13 };
  ///   assert_eq!( got, exp );
  ///
  ///   let got : Struct1 = make!( 1, 3 );
  ///   let exp = Struct1{ a : 1, b : 3 };
  ///   assert_eq!( got, exp );
  /// }
  ///
  /// ```
  ///
  /// ### To add to your project
  ///
  /// ``` shell
  /// cargo add type_constructor
  /// ```
  ///
  /// ## Try out from the repository
  ///
  /// ``` shell test
  /// git clone https://github.com/Wandalen/wTools
  /// cd wTools
  /// cd sample/rust/type_constructor_trivial_sample
  /// cargo run
  /// ```

  #[ macro_export ]
  macro_rules! make
  {

    (
      $(,)?
    )
    =>
    {
      $crate::Make0::make_0();
    };

    (
      $Arg1 : expr $(,)?
    )
    =>
    {
      $crate::Make1::make_1( $Arg1 );
    };

    (
      $Arg1 : expr, $Arg2 : expr $(,)?
    )
    =>
    {
      $crate::Make2::make_2( $Arg1, $Arg2 );
    };

    (
      $Arg1 : expr, $Arg2 : expr, $Arg3 : expr $(,)?
    )
    =>
    {
      $crate::Make3::make_3( $Arg1, $Arg2, $Arg3 );
    };

    (
      $( $Rest : tt )+
    )
    =>
    {
      compile_error!
      (
        concat!
        (
          "Variadic constructor supports up to 3 arguments.\n",
          "Open an issue if you need more.\n",
          "You passed:\n",
          stringify!
          (
            make!( $( $Rest )+ )
          )
        )
      );
    };

  }

  pub use make;
}

/// Protected namespace of the module.
pub mod protected
{
  #[ doc( inline ) ]
  pub use super::orphan::*;
}

#[ doc( inline ) ]
pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  #[ doc( inline ) ]
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  #[ doc( inline ) ]
  pub use super::prelude::*;
}

#[ doc( inline ) ]
pub use exposed::*;

/// Prelude to use essentials: `use my_module::prelude::*`.
pub mod prelude
{
  #[ cfg( feature = "make" ) ]
  #[ doc( inline ) ]
  pub use super::private::
  {

    Make0,
    Make1,
    Make2,
    Make3,

    make,

  };
}
