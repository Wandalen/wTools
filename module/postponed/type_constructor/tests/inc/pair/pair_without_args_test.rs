use type_constructor::prelude::*;

fn main()
{
  types!( pair Empty : <  > );
}
