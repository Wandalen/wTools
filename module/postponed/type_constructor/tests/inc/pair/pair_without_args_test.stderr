error: no rules expected the token `>`
  --> tests/dt/type_constructor/pair/pair_without_args_test.rs:5:27
   |
5  |   types!( pair Empty : <  > );
   |                           ^ no rules expected this token in macro call
   |
note: while trying to match meta-variable `$ParamName1:ident`
  --> rust/impl/dt/type_constructor/pair.rs
   |
   |         $ParamName1 : ident $( : $ParamTy1x1 : ident $( :: $ParamTy1xN : ident )* $( + $ParamTy1x2 : path )* )?,
   |         ^^^^^^^^^^^^^^^^^^^
